File: frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { Observable, throwError } from 'rxjs';
import { ApiService } from './api.service';
import { User, UserRequest, PageResponse } from '../models';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private endpoint = '/users';

  constructor(private api: ApiService) {}

  getUsers(page: number = 0, size: number = 50, sortBy?: string, sortDir?: string): Observable<PageResponse<User>> {
    const params: any = {};
    if (sortBy) params.sortBy = sortBy;
    if (sortDir) params.sortDir = sortDir;
    return this.api.getPagedData<User>(this.endpoint, page, size, params);
  }

  searchUsers(params: {
    name?: string;
    email?: string;
    employeeId?: string;
    department?: string;
    role?: string;
    status?: string;
    sortBy?: string;
    sortDir?: string;
  }, page: number = 0, size: number = 50): Observable<PageResponse<User>> {
    const searchParams = new URLSearchParams();
    searchParams.append('page', page.toString());
    searchParams.append('size', size.toString());
    
    Object.keys(params).forEach(key => {
      const value = params[key as keyof typeof params];
      if (value !== null && value !== undefined && value !== '') {
        searchParams.append(key, value.toString());
      }
    });
    
    return this.api.get<PageResponse<User>>(`${this.endpoint}/search?${searchParams.toString()}`);
  }

  getUserById(id: number): Observable<User> {
    return this.api.get<User>(`${this.endpoint}/${id}`);
  }

  getUserByEmployeeId(employeeId: string): Observable<User> {
    if (!employeeId?.trim()) {
      return throwError(() => new Error('Employee ID is required'));
    }
    return this.api.get<User>(`${this.endpoint}/employee/${employeeId.trim()}`);
  }

  getUserByEmail(email: string): Observable<User> {
    if (!email?.trim() || !email.includes('@')) {
      return throwError(() => new Error('Valid email is required'));
    }
    return this.api.get<User>(`${this.endpoint}/email/${email.trim()}`);
  }

  getUsersByRole(role: string, page: number = 0, size: number = 50): Observable<PageResponse<User>> {
    return this.api.getPagedData<User>(`${this.endpoint}/role/${role}`, page, size);
  }

  getUsersByDepartment(department: string, page: number = 0, size: number = 50): Observable<PageResponse<User>> {
    return this.api.getPagedData<User>(`${this.endpoint}/department/${department}`, page, size);
  }

  getActiveUsers(page: number = 0, size: number = 50): Observable<PageResponse<User>> {
    return this.api.getPagedData<User>(`${this.endpoint}/status/ACTIVE`, page, size);
  }

  createUser(user: UserRequest): Observable<User> {
    return this.api.post<User>(this.endpoint, user);
  }

  updateUser(id: number, user: UserRequest): Observable<User> {
    return this.api.put<User>(`${this.endpoint}/${id}`, user);
  }

  deleteUser(id: number): Observable<void> {
    return this.api.delete<void>(`${this.endpoint}/${id}`);
  }

  changePassword(userId: number, currentPassword: string, newPassword: string): Observable<void> {
    const payload = { currentPassword, newPassword };
    return this.api.put<void>(`${this.endpoint}/${userId}/change-password`, payload);
  }

  adminResetPassword(userId: number, newPassword: string): Observable<void> {
    const payload = { newPassword };
    return this.api.put<void>(`${this.endpoint}/${userId}/admin-reset-password`, payload);
  }

  updateUserStatus(userId: number, status: string): Observable<User> {
    return this.api.put<User>(`${this.endpoint}/${userId}/status`, { status });
  }
}

File: backend/src/main/java/com/assetdesk/service/UserService.java
package com.assetdesk.service;

import com.assetdesk.dto.user.UserRequestDTO;
import com.assetdesk.dto.user.UserUpdateDTO;
import com.assetdesk.dto.user.UserResponseDTO;
import com.assetdesk.dto.user.ChangePasswordDTO;
import com.assetdesk.dto.user.AdminPasswordResetDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface UserService {
    
    UserResponseDTO createUser(UserRequestDTO userRequestDTO);
    UserResponseDTO getUserById(Long id);
    UserResponseDTO getUserByEmployeeId(String employeeId);
    UserResponseDTO getUserByEmail(String email);
    Page<UserResponseDTO> getAllUsers(Pageable pageable);
    Page<UserResponseDTO> getUsersByRole(String role, Pageable pageable);
    Page<UserResponseDTO> getUsersByDepartment(String department, Pageable pageable);
    Page<UserResponseDTO> getUsersByStatus(String status, Pageable pageable);
    UserResponseDTO updateUser(Long id, UserUpdateDTO userUpdateDTO);
    void deleteUser(Long id);
    void changePassword(Long userId, ChangePasswordDTO changePasswordDTO);
    void adminResetPassword(Long userId, AdminPasswordResetDTO adminPasswordResetDTO);
    Page<UserResponseDTO> searchUsers(String name, String email, String employeeId, 
        String department, String role, String status, Pageable pageable);
}